Recently, a newly sequencing technique becomes much more polular among _de novo_ genome assembly research, optical map, which could greatly improve the length of scaffold. You could read the _nature biotechnology_ paper ([Genome mapping on nanochannel arrays for structural variation analysis and sequence assembly](https://www.nature.com/articles/nbt.2303)) for more details about the nanochannel array technique.
You will get **.bnx** files for future analysis and there are mainly three softwares, [Bionano Genomics IrysView software](https://bionanogenomics.com/support-page/irysview/), [KSU scripts](https://github.com/i5K-KINBRE-script-share/Irys-scaffolding) and [runBNG](https://github.com/AppliedBioinformatics/runBNG). I have tried all these three softwares and greatly prefer the runBNG software. 
* Bionano Genomics IrysView supports **Windows OS 64-bit systems** only, and you could also explore the protocol and guide by its [website](https://bionanogenomics.com/support-page/irysview/) which provides the practice samples, learning PDFs and videos.
* KSU scripts provide three pipelines (i) Sewing Machine pipeline: iteratively super scaffold genome FASTA files with BioNano genome maps using stitch.pl, (ii) "Raw data-to-finished assembly and assembly analysis" pipeline for BioNano molecule maps with a sequence-based genome FASTA, and (iii) "Raw data-to-finished de novo assembly and assembly analysis" pipeline for BioNano molecule maps.
